src/QueryM.purs

```diff
@@ -813,70 +772,79 @@ listeners :: forall (listeners :: Type). WebSocket listeners -> listeners
 listeners (WebSocket _ ls) = ls
 
 -- interface required for adding/removing listeners
-type DatumCacheListeners = ListenerSet DcWsp.JsonWspResponse
+type DatumCacheListeners = ListenerSet DatumCacheRequest DcWsp.JsonWspResponse
+
+type PendingRequests (request :: Type) = Ref (Map ListenerId RequestBody)
 
 type OgmiosListeners =
-  { utxo :: ListenerSet Ogmios.UtxoQR
-  , chainTip :: ListenerSet Ogmios.ChainTipQR
-  , submit :: ListenerSet Ogmios.SubmitTxR
-  , evaluate :: ListenerSet Ogmios.TxEvaluationResult
+  { utxo :: ListenerSet Ogmios.OgmiosAddress Ogmios.UtxoQR
+  , chainTip :: ListenerSet Unit Ogmios.ChainTipQR
+  , submit :: ListenerSet { txCbor :: ByteArray } Ogmios.SubmitTxR
+  , evaluate :: ListenerSet { txCbor :: ByteArray } Ogmios.TxEvaluationResult
   }
 
+type RequestBody = String
+
 -- convenience type for adding additional query types later
-type ListenerSet a =
-  { addMessageListener :: String -> (a -> Effect Unit) -> Effect Unit
-  , removeMessageListener :: String -> Effect Unit
-  , dispatchIdMap :: DispatchIdMap a
+type ListenerSet (request :: Type) (response :: Type) =
+  { addMessageListener :: ListenerId -> (response -> Effect Unit) -> Effect Unit
+  , removeMessageListener :: ListenerId -> Effect Unit
+  -- ^ Removes ID from dispatch map and pending requests queue.
+  , addRequest :: ListenerId -> RequestBody -> Effect Unit
+  -- ^ Saves request body until the request is fulfilled. The body is used
+  --  to replay requests in case of a WebSocket failure.
   }
 
 -- we manipluate closures to make the DispatchIdMap updateable using these
 -- methods, this can be picked up by a query or cancellation function
-mkListenerSet :: forall (a :: Type). DispatchIdMap a -> ListenerSet a
-mkListenerSet dim =
+mkListenerSet
+  :: forall (request :: Type) (response :: Type)
+   . DispatchIdMap response
+  -> PendingRequests request
+  -> ListenerSet request response
+mkListenerSet dim pr =
   { addMessageListener:
-      \id -> \func -> do
-        idMap <- Ref.read dim
-        Ref.write (MultiMap.insert id func idMap) dim
+      \id func -> do
+        Ref.modify_ (MultiMap.insert id func) dim
   , removeMessageListener:
       \id -> do
-        idMap <- Ref.read dim
-        Ref.write (MultiMap.delete id idMap) dim
-  , dispatchIdMap: dim
+        Ref.modify_ (MultiMap.delete id) dim
+        Ref.modify_ (Map.delete id) pr
+  , addRequest:
+      \id req ->
+        Ref.modify_ (Map.insert id req) pr
   }
 
-removeAllListeners
-  :: forall (a :: Type). LogLevel -> DispatchIdMap a -> Effect Unit
-removeAllListeners lvl dim = do
-  logString lvl Error "error hit, removing all listeners"
-  Ref.write MultiMap.empty dim
-
 -- TODO after ogmios-datum-cache implements reflection this could be generalized to make request for the cache as well
```

Comment:
I know this isn't part of the issues that this PR addresses, but do you think it would be possible to address this TODO as well? (Since you're already changing the internals of the jsonwsp stuff.) AFAIK @adamczykm made a PR against ODC that implemented reflection 

If you don't want to include it in this PR, we should open an issue for this and add the URL